plugins/inputs/suricata/suricata.go

```diff
@@ -0,0 +1,215 @@
+}
+
+const sampleConfig = `
+  ## Data sink for Suricata stats log
+  # This is expected to be a filename of a
+  # unix socket to be created for listening.
+  # Will be overwritten if a socket or file
+  # with that name already exists.
+  source = "/tmp/suricata-stats.sock"
+`
+
+// SampleConfig returns a sample TOML section to illustrate configuration
+// options.
+func (s *Suricata) SampleConfig() string {
+	return sampleConfig
+}
+
+// Start initiates background collection of JSON data from the socket
+// provided to Suricata.
+func (s *Suricata) Start(acc telegraf.Accumulator) error {
+	var err error
+	s.Lock()
+	defer s.Unlock()
+	if s.InputListener == nil {
+		os.Remove(s.Source)
+		s.InputListener, err = net.ListenUnix("unix", &net.UnixAddr{
+			Name: s.Source,
+			Net:  "unix",
+		})
+		if err != nil {
+			return err
+		}
+		s.CloseChan = make(chan bool)
+		s.InputListener.SetUnlinkOnClose(true)
+		go s.handleServerConnection(acc)
+	}
+	return nil
+}
+
+// Stop causes the plugin to cease collection JSON data from the socket provided
+// to Suricata.
+func (s *Suricata) Stop() {
+	s.Lock()
+	defer s.Unlock()
+	if s.CloseChan != nil {
+		s.InputListener.Close()
+		s.ClosedChan = make(chan bool)
+		close(s.CloseChan)
+		<-s.ClosedChan
+		s.CloseChan = nil
+	}
+}
+
+func (s *Suricata) handleServerConnection(acc telegraf.Accumulator) {
+	var err error
+	for {
+		select {
+		case <-s.CloseChan:
+			s.InputListener = nil
+			close(s.ClosedChan)
+			return
+		default:
+			var conn net.Conn
+			conn, err = s.InputListener.Accept()
+			if err != nil {
+				continue
+			}
+			reader := bufio.NewReaderSize(conn, 10485760)
+		out:
+			for {
+				select {
+				case <-s.CloseChan:
+					conn.Close()
+					s.InputListener.Close()
+					s.InputListener = nil
+					close(s.ClosedChan)
+					return
+				default:
+					line, isPrefix, rerr := reader.ReadLine()
+					if rerr == nil {
+						if isPrefix {
+							acc.AddError(errors.New("incomplete line read from input"))
+							continue
+						} else {
+							s.parse(acc, line)
+						}
+					} else if rerr == io.EOF {
+						break out
+					}
+				}
+			}
+		}
+	}
+}
+
+func splitAtSingleDot(in string) []string {
+	res := singleDotRegexp.FindAllStringIndex(in, -1)
+	if res == nil {
+		return []string{in}
+	}
+	ret := make([]string, 0)
+	startpos := 0
+	for _, v := range res {
+		ret = append(ret, in[startpos:v[0]+1])
+		startpos = v[1] - 1
+	}
+	return append(ret, in[startpos:])
+}
+
+func addPercentage(thread string, val1 string, val2 string, valt string,
+	fields map[string](map[string]interface{})) {
+	if _, ok := fields[thread][val1]; ok {
+		if _, ok := fields[thread][val2]; ok {
+			f1 := fields[thread][val1].(float64)
+			f2 := fields[thread][val2].(float64)
+			fields[thread][valt] = f2 / (f1 + f2)
+		}
+	}
+}
+
+func (s *Suricata) parse(acc telegraf.Accumulator, sjson []byte) {
+	if len(sjson) == 0 {
+		return
+	}
+
+	out, err := gojsonexplode.Explodejsonstr(string(sjson), ".")
```

Comment:
Unfortunately there are `_` in the field names that are present in the stats JSON, so using that as a delimiter would be problematic.
Do you expect a change in your implementation to be more flexible with the delimiter? If so I'd be happy to drop this dependency as soon as it is possible.

Reply:
> Do you expect a change in your implementation to be more flexible with the delimiter?

No planning to make a change, but we no longer allow `.` in field names anyway.  The reasons being:
- Some outputs don't support it (prometheus)
- Acts funny in other outputs that use dot delimited values (graphite)
- Needs escaped in others (influxdb)
- Style consistency
- Fields are meant to be a single "layer", we try to use tags for what would traditionally be hierarchical organization.

Reply:
That would be a hierarchy multiple levels deep, sometimes with only one actual field. I also think that it would be quite counterintuitive to the Suricata community as these counters are typically referred to as a single key-value pairs. For instance, here's the tabular output that Suricata produces locally:
```
------------------------------------------------------------------------------------
Date: 8/9/2019 -- 08:55:22 (uptime: 3d, 15h 27m 05s)
------------------------------------------------------------------------------------
Counter                                       | TM Name                   | Value
------------------------------------------------------------------------------------
capture.kernel_packets                        | Total                     | 122137797483
capture.kernel_drops                          | Total                     | 49263200341
decoder.pkts                                  | Total                     | 72874365061
decoder.bytes                                 | Total                     | 53804490787164
decoder.invalid                               | Total                     | 506
decoder.ipv4                                  | Total                     | 72877053154
decoder.ipv6                                  | Total                     | 59884
decoder.ethernet                              | Total                     | 72874365061
decoder.tcp                                   | Total                     | 42319193466
decoder.udp                                   | Total                     | 30413560427
decoder.icmpv4                                | Total                     | 74834392
decoder.icmpv6                                | Total                     | 289
decoder.teredo                                | Total                     | 59884
decoder.avg_pkt_size                          | Total                     | 738
decoder.max_pkt_size                          | Total                     | 1530
flow.tcp                                      | Total                     | 589114037
flow.udp                                      | Total                     | 139686591
flow.icmpv4                                   | Total                     | 17713676
flow.icmpv6                                   | Total                     | 91
defrag.ipv4.fragments                         | Total                     | 16097095
defrag.ipv4.reassembled                       | Total                     | 3324169
decoder.event.ipv4.opt_pad_required           | Total                     | 102113
decoder.event.icmpv4.unknown_code             | Total                     | 37768
decoder.event.icmpv4.ipv4_unknown_ver         | Total                     | 502
decoder.event.ipv6.unknown_next_header        | Total                     | 2
decoder.event.tcp.opt_invalid_len             | Total                     | 255
decoder.event.udp.hlen_invalid                | Total                     | 3
...
```
I would be in favour of keeping the field keys as they are. We are actually using these at the moment to populate an InfluxDB without extra escaping  or processing. Alternatively, I'd propose a configuration option to switch delimiters, with `_` being the default and we could override it to `.` if desired. OK?

Reply:
I implemented this option and also dropped the dependency on the external library by adding a few lines of tailored code for the expected input.

Reply:
That sounds like it should work.

If your curious, here is the issue I mentioned in InfluxQL. It only affects measurements and only comes up if you have a retention policy/database with the same name:
```
> use telegraf.autogen
> insert autogen.foo value=42
> insert foo value=43
> select * from autogen.foo
name: foo
time                value
----                -----
1565376105031754812 43
> select * from "autogen.foo"
name: autogen.foo
time                value
----                -----
1565376102698991940 42
```

```
> insert autogen.bar value=42
> select * from autogen.bar
> select * from "autogen.bar"
name: autogen.bar
time                value
----                -----
1565376587386068626 42
```

Reply:
Ah I see. Not a problem for our use case but still OK to default to an unproblematic separator.